### PR TITLE
Test that UTxO tables are updated correctly on era transitions

### LINKED ISSUE #4043: Cardano eras transitions tests
- [x] Write tests for all the Cardano transitions. 
    - [x] Byron to Shelley
    - [x] Shelley to Allegra
    - [x] Allegra to Mary
    - [x] Mary to Alonzo 
    - [x] Alonzo to Babbage
- [x] Add to the `AddingAnEra.md` document instructions to add a new era translation test when adding a new era.

If we can generate arbitrary ledger states of each Cardano era we can write property tests. Otherwise, we'll have to resort to unit tests.

# Byron to Shelley 
Ensure that *all the UTxO* is transferred from in-memory Byron state (which has no tables) to the ledger tables of the Shelley state. The result must consist only of insertions (which are present in the resulting diffs).

# Shelley to Allegra
Ensure that the *avvm* addresses appear as deletions in the resulting diffs.

# All remaining transitions
Ensure that the resulting diff is empty, which means no modification to the ledger tables will be made from one era to the next.

### PR DESCRIPTION
**Pull request by bartfrenk**
# Description

Test that UTxO tables are computed correctly on era transitions, for all era transitions.

Closes #4043.

# Checklist

- Branch
    - [x] Commit sequence broadly makes sense
    - [x] Commits have useful messages
    - [x] New tests are added if needed and existing tests are updated
- Pull Request
    - [x] Self-reviewed the diff
    - [x] Useful pull request description at least containing the following information:
      - What does this PR change?
      - Why these changes were needed?
      - How does this affect downstream repositories and/or end-users?
      - Which ticket does this PR close (if any)? If it does, is it [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)?
    - [x] Reviewer requested

### CODE REVIEW

File: ouroboros-consensus-byron-test/src/Test/Consensus/Byron/Generators.hs [hunk 1]
```diff
@@ -272,6 +278,28 @@ instance Arbitrary ByronTransition where
 instance Arbitrary (LedgerState ByronBlock EmptyMK) where
   arbitrary = ByronLedgerState <$> arbitrary <*> arbitrary <*> arbitrary
 
+-- | Generator for a Byron ledger state in which the tip of the ledger given by
+-- `byronLedgerTipBlockNo` matches that derived from the `ChainValidationState` in
```
Comment (dnadales) [inline ouroboros-consensus-byron-test/src/Test/Consensus/Byron/Generators.hs:282]: We should explain what *matches* means here. Maybe something like: 

> block number of the ledger tip is consistent with the chain validation state, this is, if there is no previous block, then the ledger tip will be origin

File: ouroboros-consensus-byron-test/src/Test/Consensus/Byron/Generators.hs [hunk 2]
```diff
@@ -272,6 +278,28 @@ instance Arbitrary ByronTransition where
...
+-- `byronLedgerState`.
+genByronLedgerState :: Gen (LedgerState ByronBlock EmptyMK)
```
Comment (dnadales) [inline ouroboros-consensus-byron-test/src/Test/Consensus/Byron/Generators.hs:284]: While it makes sense that the tip block number matches the chain state, I wonder if we should generate ledger states in which this is not the case. Don't we want to test our system under these circumstances?

File: ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+  ) where
+
+import qualified Data.ListMap as ListMap
+import           Data.Map.Strict (Map)
+import qualified Data.Map.Strict as Map
+import           Test.QuickCheck
+import           Test.Tasty
+import           Test.Tasty.QuickCheck
+
+import qualified Cardano.Chain.Block as Byron
```
Comment (dnadales) [inline ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs:18]: Just to check: does this group of imports correspond to `cardano-ledger`?

File: ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+import           Test.Consensus.Cardano.MockCrypto (MockCryptoCompatByron)
+import           Test.Consensus.Shelley.Generators ()
+import           Test.Consensus.Shelley.MockCrypto
+import           Test.Util.Time (dawnOfTime)
+
+import qualified Cardano.Ledger.Block as Core
+import qualified Cardano.Ledger.Core as Core
+import qualified Cardano.Ledger.Era as Core
+
+type Crypto = MockCryptoCompatByron
```
Comment (dnadales) [inline ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs:69]: I wonder if we should remove this level of indirection.

File: ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+type Era    = ShelleyEra Crypto
+type Proto  = TPraos Crypto
+
+data TestSetup = TestSetup {
```
Comment (dnadales) [inline ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs:73]: Are we going to have one test setup per-each era transition?

File: ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+    TestSetup <$> genByronLedgerConfig
+              <*> pure fixedShelleyLedgerConfig
+              <*> genByronLedgerState
+
+tests :: TestTree
+tests = testGroup "ByronToShelley" [
+    testProperty "test" prop_byronUtxosMatchThoseInShelleyTable
+  ]
+
+prop_byronUtxosMatchThoseInShelleyTable :: TestSetup -> Property
```
Comment (dnadales) [inline ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs:95]: I wonder if we should comment the background of this test: why do we need it?

File: ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+prop_byronUtxosMatchThoseInShelleyTable TestSetup {..} =
```
Comment (dnadales) [inline ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs:96]: Should we avoid wildcards, so that it is clearer where each identified is declared?

File: ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs [hunk 8]
```diff
@@ -0,0 +1,226 @@
...
+
+fixedShelleyLedgerConfig :: Core.TranslationContext era ~ ()
+                         => ShelleyLedgerConfig era
+fixedShelleyLedgerConfig = mkShelleyLedgerConfig
+      shelleyGenesis
+      ()
+      (fixedEpochInfo (sgEpochLength shelleyGenesis) tpraosSlotLength)
+      (MaxMajorProtVer 1000)
+
+shelleyGenesis :: ShelleyGenesis era
```
Comment (dnadales) [inline ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs:206]: If we use a fixed Shelley genesis we should justify why. I think in this case the justification might be that we are interested in the initial UTxO set, which is not affected by the genesis config. But we need to confirm this.

File: ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+  -> LedgerState (ShelleyBlock Proto (ShelleyEra Crypto)) EmptyMK
+  -> LedgerState (ShelleyBlock Proto (AllegraEra Crypto)) DiffMK
+translateLedgerStateShelleyToAllegra shelleyConfig allegraConfig epochNo shelleyLedgerState =
+  let translateLedgerState = shelleyToAllegraTranslateLedgerState shelleyConfig allegraConfig
+  in translateLedgerStateWith translateLedgerState epochNo shelleyLedgerState
+
+genShelleyLedgerState :: CanMock proto era => Gen (LedgerState (ShelleyBlock proto era) EmptyMK)
+genShelleyLedgerState = arbitrary
+
+run :: IO ()
```
Comment (dnadales) [inline ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs:169]: Is this temporary?

File: ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+
+tests :: TestTree
+tests = testGroup "ByronToShelley" [
+    testProperty "test" prop_byronUtxosMatchThoseInShelleyTable
+  ]
+
+prop_byronUtxosMatchThoseInShelleyTable :: TestSetup -> Property
+prop_byronUtxosMatchThoseInShelleyTable TestSetup {..} =
+  let
+    epochNo = EpochNo 10
```
Comment (dnadales) [inline ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs:98]: What is the effect of the epoch number in this property, and/or in the ledger state translation?

File: ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+    gen :: Gen (LedgerState (ShelleyBlock Proto (AllegraEra Crypto)) DiffMK)
+    gen = do
+      let shelleyC = fixedShelleyLedgerConfig
+      let allegraC = fixedShelleyLedgerConfig
+      state <- genShelleyLedgerState
+      pure $ translateLedgerStateShelleyToAllegra shelleyC allegraC epochNo state
+    epochNo :: EpochNo
+    epochNo = EpochNo 10
+
+translateToShelleyLedgerState
```
Comment (dnadales) [inline ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs:181]: What about?

```suggestion
translateByronToShelleyLedgerState
```

File: ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+  :: LedgerConfig ByronBlock
+  -> LedgerConfig (ShelleyBlock Proto Era)
+  -> EpochNo
+  -> LedgerState ByronBlock EmptyMK
+  -> LedgerState (ShelleyBlock Proto Era) DiffMK
+translateToShelleyLedgerState byronConfig shelleyConfig epochNo byronLedgerState =
```
Comment (dnadales) [inline ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs:187]: Would it make sense to make these definitions local to their use site?

**Review by dnadales — COMMENTED** [review]
Looks good so far.

File: ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs (hunk 5 above)
Comment (bartfrenk) [reply ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs:73]: Yes

File: ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs (hunk 4 above)
Comment (bartfrenk) [reply ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs:86]: I think it is unnecessary indirection, but it reduces noise in type signatures. I am impartial to this one.

File: ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs (hunk 3 above)
Comment (bartfrenk) [reply ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs:27]: I have checked the imports, although from the style guide is not completely clear to me how to order imports (I'd prefer it to be simpler, but that is another discussion).

File: ouroboros-consensus-byron-test/src/Test/Consensus/Byron/Generators.hs (hunk 2 above)
Comment (bartfrenk) [reply ouroboros-consensus-byron-test/src/Test/Consensus/Byron/Generators.hs:284]: The translation then fails; my assumption that "invalid" ledger states are only due to programming errors, in which case I don't think we should test the behaviour, but that assumption might be wrong.

File: ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs (hunk 4 above)
Comment (dnadales) [reply ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs:86]: I'd prefer skipping the extra level of indirection, but that's subjective. @Jasagredo preferences?

File: ouroboros-consensus-byron-test/src/Test/Consensus/Byron/Generators.hs [hunk 13]
```diff
@@ -272,6 +278,28 @@ instance Arbitrary ByronTransition where
...
+genByronLedgerState = do
+  chainValidationState <- arbitrary
+  ledgerTransition <- arbitrary
+  ledgerTipBlockNo <- genLedgerTipBlockNo chainValidationState
+  pure $ ByronLedgerState {
+      byronLedgerTipBlockNo = ledgerTipBlockNo
```
Comment (dnadales) [inline ouroboros-consensus-byron-test/src/Test/Consensus/Byron/Generators.hs:290]: I think we try to align this in other parts of the code base.

File: ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs (hunk 3 above)
Comment (dnadales) [reply ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs:27]: So do they correspond to `cardano-ledger`? :smile: To put it differently, why do we have this group of imports. My assumption was that you grouped the import in this way because they all belong to the same package. I'm not objecting to this grouping ;)

File: ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs [hunk 14]
```diff
@@ -0,0 +1,320 @@
...
+import           Ouroboros.Consensus.Byron.Ledger (ByronBlock, byronLedgerState)
+
+import           Ouroboros.Consensus.Shelley.Eras
+import           Ouroboros.Consensus.Shelley.HFEras ()
+import           Ouroboros.Consensus.Shelley.Ledger (ShelleyBlock,
+                     ShelleyLedgerConfig, mkShelleyLedgerConfig,
+                     shelleyLedgerState, shelleyLedgerTables, shelleyUTxOTable)
+import           Ouroboros.Consensus.Shelley.Ledger.SupportsProtocol ()
+
+-- TODO: Not sure where to put these
```
Comment (dnadales) [inline ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs:74]: I think they're ok here. @Jasagredo ?

File: ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs [hunk 15]
```diff
@@ -0,0 +1,320 @@
...
+
+-- Definitions to make the signatures a bit less unwieldy
+type Crypto = MockCryptoCompatByron
+type Proto  = TPraos Crypto
+
+tests :: TestTree
+tests = testGroup "UpdateTablesOnEraTransition" $ toTestTrees predicates tls
+  where
+    tls = translateLedgerState hardForkEraTranslation
+    predicates :: InPairs TranslationPredicate (CardanoEras Crypto)
```
Comment (dnadales) [inline ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs:93]: Why not define a `[TestTree]` directly? Is `toTestTrees` necessary?

File: ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs [hunk 16]
```diff
@@ -0,0 +1,320 @@
...
+  , tsDestLedgerConfig :: LedgerConfig dest
+  , tsSrcLedgerState   :: LedgerState src EmptyMK
+  , tsEpochNo          :: EpochNo
+}
+
+deriving instance ( Show (LedgerConfig src)
+                  , Show (LedgerConfig dest)
+                  , Show (LedgerState src EmptyMK)) => Show (TestSetup src dest)
+
+-- todo: Useful to merge some of these instances?
```
Comment (dnadales) [inline ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs:232]: I don't see how we could merge the Byron to Shelley instances. I'm not sure about the others. Do you know what prevents us from defining a:

```haskell 
instance Arbitrary (TestSetup (ShelleyBlock p era0) (ShelleyBlock p era1))
```

File: ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs [hunk 17]
```diff
@@ -0,0 +1,320 @@
...
+-- transitions does not depend on the configurations of any of the ledgers involved..
+-- TODO: Check this
+fixedShelleyLedgerConfig :: Core.TranslationContext era -> ShelleyLedgerConfig era
+fixedShelleyLedgerConfig translationContext = mkShelleyLedgerConfig
+      shelleyGenesis
+      translationContext
+      (fixedEpochInfo (sgEpochLength shelleyGenesis) (slotLengthFromSec 2))
+      (MaxMajorProtVer 1000)
+  where
+    shelleyGenesis = ShelleyGenesis {
```
Comment (dnadales) [inline ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs:286]: We should align these.

File: ouroboros-consensus/src/Ouroboros/Consensus/Util.hs [hunk 18]
```diff
@@ -346,6 +350,10 @@ allDisjoint = go Set.empty
     go _   []       = True
     go acc (xs:xss) = Set.disjoint acc xs && go (Set.union acc xs) xss
 
+-- | Map over keys and values
```
Comment (dnadales) [inline ouroboros-consensus/src/Ouroboros/Consensus/Util.hs:353]: Why is this called a `dimap` and not a `bimap`? My guess: `dimap` does not satisfy the bifunctor laws (I haven't checked).

**Review by dnadales — APPROVED** [review]

**Comment by dnadales** [comment]
I forgot to add these remaining tasks:

- [ ] Add to the AddingAnEra.md document instructions to add a new era translation test when adding a new era.
- [ ] Use quickcheck's [`cover`](https://www.stackage.org/haddock/lts-19.29/QuickCheck-2.14.2/Test-QuickCheck.html#v:cover) to check that we're testing important cases.

File: ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs [hunk 19]
```diff
@@ -0,0 +1,320 @@
...
+}
+
+toTestTrees
+  :: (AllPairs TestSetup Arbitrary xs, AllPairs TestSetup Show xs)
+  => InPairs TranslationPredicate xs
+  -> InPairs (RequiringBoth WrapLedgerConfig TranslateLedgerState) xs
+  -> [TestTree]
+toTestTrees predicates translations = mkTestTrees $ zipInPairs predicates translations
+  where
+    fn
```
Comment (Jasagredo) [inline ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs:177]: I would avoid naming this `fn` because we are heavily into `generics-sop` in these hard fork tests and there is actually a function named like that in SOP: https://hackage.haskell.org/package/generics-sop-0.5.1.2/docs/Generics-SOP.html#v:fn

File: ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs (hunk 4 above)
Comment (bartfrenk) [reply ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs:86]: I agree with both, but I also tend to read files as @Jasagredo. So just to close the discussion I am keeping it as it is.

File: ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs (hunk 7 above)
Comment (bartfrenk) [reply ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs]: I have no preference. The prefixes in this case make it clear.

File: ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs (hunk 15 above)
Comment (bartfrenk) [reply ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs:95]: Well, I couldn't do it.

File: ouroboros-consensus/src/Ouroboros/Consensus/Util.hs (hunk 18 above)
Comment (bartfrenk) [reply ouroboros-consensus/src/Ouroboros/Consensus/Util.hs:353]: It is a profunctor I think, in the sense that it contravariant in the first argument, covariant in the latter. The profunctor map in https://hackage.haskell.org/package/profunctors-5.6.2/docs/Data-Profunctor.html is called `dimap`.

File: ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Util/InPairs.hs [hunk 20]
```diff
@@ -120,3 +123,15 @@ requiringBoth = flip go
     go :: InPairs (RequiringBoth h f) xs -> NP h xs -> InPairs f xs
     go PNil         _              = PNil
     go (PCons f fs) (x :* y :* zs) = PCons (provideBoth f x y) (go fs (y :* zs))
+
+{-------------------------------------------------------------------------------
+  Constraints
+-------------------------------------------------------------------------------}
+
+type family AllPairs
```
[the remark this answers is not in the document]
Comment (bartfrenk) [reply ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Util/InPairs.hs:131]: This is adapted from Maguire's Thinking with types. I can kind-of see how to do it with classes, but I like the terseness of this definition, and I think, but do not know, the class-one will be more convoluted.

File: ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs (hunk 15 above)
Comment (bartfrenk) [reply ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/Translation.hs:95]: To define the properties I need to have access to the appropriate `TranslateEra x y` values. I chose the direction of zipping the predicates with what is loosely, `InPairs TranslateEra (CardanoEras c)`.

**Comment by dnadales** [comment]
We should also test with empty UTxO, shouldn't we? 

```text
      +++ OK, passed 100 tests (100% UTxO set is not empty).
```

It seems we aren't at the moment.
